### ISSUE TITLE
Degrading performance when generating many functions

### ISSUE THREAD

**Issue by rcnlee** [JuliaLang/julia#18446]
I am using Julia for Genetic Programming, so my problem inherently involves declaring new functions at runtime using `@eval`.  However, I am noticing that the performance penalty is not fixed --- it seems to grow with iterations.

Please consider the following example:

``` julia
module A
export foo, bar
export to_function

f(x) = 0
g(x) = x + 1

function to_function(expr)
    @eval f(x) = $expr
    f
end
function foo(expr)
    f = to_function(expr)
    for x=1:1000
       f(x)
    end
    f
end
function bar(expr)
    f = to_function(expr)
    for x=1:1000
        g(x)
    end
    f
end
end #module

using A

@time expr = :(x+1)
@time to_function(expr)
A.f(1)
foo(expr)
bar(expr)

for i=1:1000000
    println("$i: foo-bar")
    i2 = rand(1:i)
    i3 = rand(1:i)
    i4 = rand(1:i)
    i5 = rand(1:i)
    i6 = rand(1:i)
    i7 = rand(1:i)
    i8 = rand(1:i)
    expr = :(x+$i+$i2+$i3+$i4+$i5+$i6+$i7+$i8)
    @time foo(expr)
    @time bar(expr)
end
```

The degradation is slow in this simplified example (faster in my original code), so you may need to run it for a few hundred thousand iterations.

Here's what I see:

``` julia
5: foo-bar
  0.001702 seconds (1.76 k allocations: 52.834 KB)
  0.000386 seconds (36 allocations: 2.298 KB)
6: foo-bar
  0.001891 seconds (1.76 k allocations: 52.834 KB)
  0.001001 seconds (36 allocations: 2.298 KB)
7: foo-bar
  0.001895 seconds (1.77 k allocations: 53.084 KB)
  0.000392 seconds (36 allocations: 2.298 KB)
......
375578: foo-bar
  0.009440 seconds (2.26 k allocations: 60.789 KB)
  0.000498 seconds (44 allocations: 2.454 KB)
375579: foo-bar
  0.009095 seconds (2.26 k allocations: 60.789 KB)
  0.000417 seconds (44 allocations: 2.454 KB)
375580: foo-bar
  0.009767 seconds (2.26 k allocations: 60.789 KB)
  0.000446 seconds (44 allocations: 2.454 KB)
```

A few things to notice:
- The `@eval`'d version takes an order of magnitude longer and allocates a huge amount of memory compared to the regular declaration.
- Performance degrades over iterations.  By a couple hundred thousand iterations, it takes 5x longer than initially.
- Memory usage (as seen on task manager) grows with iterations.  The gc comes on periodically but memory usage never goes down.  By iteration 375000 I am 3GB+ of RAM usage.

Please help!  The performance degradation is a showstopper for my Genetic Programming code!

Thanks!

**Comment by yuyichao** [comment]
This is mostly expected.

**Comment by ufechner7** [comment]
Is there a work around?

**Comment by yuyichao** [comment]
Put parameters in a type and use branches instead of creating a new function every time.

**Comment by cstjean** [comment]
Is compiled code ever GC'ed? I don't do genetic programming, but this issue also worries me for my application.

**Comment by yuyichao** [comment]
No, but the issue is more about defining functions rather than the number of specializations though.

**Comment by rcnlee** [comment]
Thanks for the replies!

I forgot to mention that the number of inner loop iterations
```julia
for i=1:1000
f(x)
end
```
has a big impact on allocations --- larger inner loop, larger allocations.  Why is that?  Shouldn't all the performance penalties be at the time of declaration/compilation rather than scaling with point of use?

@yuyichao  The biggest showstopper for me is not the fixed penalty of compiling every time, but why it gets slower and slower.  Can you elaborate on what is accumulating and slowing things down?

Thanks,

Ritchie

**Comment by yuyichao** [comment]
There are many things that are optimized for not having unreasonable number of function definitions.

**Comment by rcnlee** [comment]
Are the function definitions being accumulated? I thought there's only one ``f(x)`` under module A and it just keeps getting replaced.

**Comment by JeffBezanson** [comment]
Similar issue: #14495

**Comment by jpsamaroo** [comment]
I'm having a similar issue to this and to #14495 (another instance of genetic programming), where I generate functions using `eval`, but where each eval'd function always has a single parameter, a `Dict{Symbol,Any}`. After a small number of iterations and a short span of time, I quickly end up with my code running 2x slower than at the start, and memory ballooning like crazy.

I'm wondering if there is any sort of workaround that doesn't just end up being "restart julia" or "don't generate lots of functions", even if it means going way out of my way to tweak things. Are there any parts of julia/LLVM that are clearly causing this behavior, and are there any remedies possible, even if it would take me a lot of time and learning effort to implement?

**Comment by vchuravy** [comment]
For the use-case of genetic programming you might not want to compile your code at all. How often are you planning on executing your code? If it is only once it might be beneficial to interpret the code instead of compiling it.

**Comment by jpsamaroo** [comment]
Sure, but that's the same as asking why anyone would ever code in C/C++, when we could all just write FORTH words. Sure, FORTH might be more "flexible" for runtime-generated routines, but it's also a pain to write words efficiently, and is still not as fast on modern CPUs compared to C/C++.

That said, if there's an interpreted language with the same typing system, multiple dispatch semantics, homoiconicity (for GP), and similar speed as Julia, that's just as easy to use, please do let me know! (Though I'll still always have a soft spot for this language).

**Comment by vchuravy** [comment]
No I meant interpreting the Julia code (not switching away from Julia), but people more knowledgeable about our current efforts might be able to help more (cc: @vtjnash)

**Comment by jpsamaroo** [comment]
I guess theoretically one could step through each Expr in the AST and calculate the intermediate result, however that's immediately less efficient if you're only mutating your "DNA" every second or so (which is the case for me, my algorithms "live" for a long period of time).

**Comment by jpsamaroo** [comment]
Given that it seems like the issue is a growing list of function references or something similar, is it at all possible to free the memory of these references and shrink this hypothetical list? I'm guessing this doesn't automatically happen, because once the function is compiled, it could be used later on to speed up execution and supposedly prevent re-compilation. However, since tests show that it's obviously not being used later on during future calls to `eval` of the exact same Expr, surely it would make sense to be able to remove the reference, even if it must be done manually by the user?

I'm not sure if that's possible, but I do know that I can initialize an array, for example, and then set the variable it references to `0`. Then when the GC runs, the previously allocated memory of the array is freed, and the julia process shrinks in size. If something similar can be done for compiled functions that the user knows is no longer necessary (even at the expense of having to re-compile later, which seems to happen anyway), then that would effectively solve this problem of generated functions accumulating and slowing things down.

**Comment by JeffBezanson** [comment]
@jpsamaroo If you'd like to help, a good start would be to do some profiling to try to find where the slowdown is. I don't think we know exactly where the problem lies when the system is stressed in this way.

**Comment by vtjnash** [comment]
Closing as duplicate of https://github.com/JuliaLang/julia/issues/14495
